### PR TITLE
Further bloom improvements

### PR DESCRIPTION
**Pull request by StarLederer**
*I am assuming I don't have to follow Bevys PR guidelines here.*

This PR Updates the fork to the version I demonstrate [here](https://github.com/bevyengine/bevy/pull/6677#issuecomment-1323731180).

I might take some time to think about how to tackle aspect ratio issues. If you want to make changes to the code in the mean time i think it is will make everyone's job easier if you merge this first (if you agree with the changes of course)

### CODE REVIEW

**Comment by JMS55** [comment]
Thanks for opening this! I'll review it in a day or two - traveling due to Thanksgiving atm. Will get to it soon though.

File: crates/bevy_core_pipeline/src/bloom/bloom.wgsl
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/bloom.wgsl]: Should be settings.intensity.

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/mod.rs]: The blend factors need an explanation both here, and in the shader, about why they're used like this.

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/mod.rs]: Why did you change this to floor, and remove the -3? 

If we keep this, it should be ` let mip_count = (min_view.log2().floor() as u32).max(1);`

File: crates/bevy_core_pipeline/src/bloom/bloom.wgsl
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/bloom.wgsl:130]: At this point, can we remove upsample_final? Seems like upsample and upsample_final both do the same thing now.

**Review by JMS55 — CHANGES_REQUESTED** [review]
image

I tried out these changes, unfortunately the results don't look that great to me. There's a very sharp ring around each sphere, instead of a more spread out blur. Do you have the same issue?

File: crates/bevy_core_pipeline/src/bloom/bloom.wgsl
Comment (StarLederer) [reply crates/bevy_core_pipeline/src/bloom/bloom.wgsl:130]: That would be ideal. I am trying to come up with a mathematical representation of our interpolations that would first, remove any need for the final upsample pass, allow more artistic control, and double bloom as a free screen blur effect (can be combined with a texture to create blurred backgrounds for game UI)

I decided to keep changes to a minimum for now that's why it is there in this PR

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (StarLederer) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: Will add that. Should i explain why we are using alpha blending or that this setup is alpha blending?

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (StarLederer) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: Round sometimes results in too many mips (one of the dimensions below 1) which causes an error on the gpu.

Removed -3 so that we get closer to the smallest mip being 1x1. We need it if our definition of bloom intensity is how likely the light is to exit at a random point of the lens. Intensity of 1 should result in the entire screen being the same color (average of the original image). The change, however, still does not result in 1x1 textures though because when one of the dimensions hits 1 we cant go further. This is not an ideal solution but it is still closer to the irl counterpart

**Comment by StarLederer** [comment]
> There's a very sharp ring around each sphere

Not exactly a fan of this either but last time i checked (not too recently) this was the same behavior as Unity (though they do have an additional parameter that helps artistically fix that). The customizable mathematical representation(that i mention in a review comment) should achieve the same level of artistic control as Unity. I thought I should leave it alone for now but your review tells me it is a good idea to do it now. I'm not great at math but I'll try to get it to at lest a prototype state soon (as i mention in a review comment, this will also make it easier to get rid of upsample_final)

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (JMS55) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: Explain how the alpha doesn't actually matter because our texture is only RGB (not RGBA), but we use it to lerp between the src and dest textures.

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (JMS55) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: Ok, sounds good to me. Then you just need to fix the u32 casting like I mentioned above.

**Comment by JMS55** [comment]
Thank you, that would be great. Yeah, I'd rather do whatever changes necessary in this PR, rather then make some changes that don't look great without further commits later. Feel free to change whatever you think should be changed, so long as you leave comments and documentation about why :)

**Comment by StarLederer** [comment]
progress update: I think i figured out a function that allows for a lot of artistic control while still being physically based. Gonna implement it later today and hopefully post a progress pic

**Comment by StarLederer** [comment]
My function did not work but i ended up implementing a different solution based on bezier curves which in theory is even better but I can't really confirm because the implementation is not finished and to properly play with it we'd need a bezier curve editor.

Which begs the question, are there any bezier curve editors made with Bevy UI that i can grab or do I have to make my own? I've never done much more than just text output with Bevy UI so making one might prove time consuming.

**Comment by StarLederer** [comment]
Other than that, I did remove a part of "final upsample" but some bits still remain because the final version uses a different texture format

**Comment by JMS55** [comment]
> Which begs the question, are there any bezier curve editors made with Bevy UI that i can grab or do I have to make my own? I've never done much more than just text output with Bevy UI so making one might prove time consuming.

Not that I know of. You can setup bevy to hot-reload shaders though (including in-engine assets like the bloom shader), which would let you tweak values live.

Code looks pretty good to me, and the bloom example looks pretty good now (at 1080p) if you bump up the intensity to around ~0.5, although it makes the scene look a bit blurry. Adding a small threshold helps with that, though. We should probably tweak the default/example settings. 

At 4k, however, there's still a lot of noticeable blocky artifacts, unfortunately. Perhaps we can do the blur pyramid multiple times (dependent on resolution)?
 
image
image

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (JMS55) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: We will also need to test this on a more complex example. The bloom example in bevy originally looked fine, but testing on other scenes revealed that we needed to do -3 to avoid artifacts. Also, going down to 1x1 may not be worth it. It's probably fine to settle on 4x4 or something, and be mostly accurate, while avoiding the extra draw calls.

**Comment by JMS55** [comment]
While you're working on bloom, do you mind adding the fix from https://github.com/bevyengine/bevy/pull/6802? We're not accounting for the viewport size correctly in the current bloom impl.

**Comment by StarLederer** [comment]
> and the bloom example looks pretty good now

Well.. the look is exactly the same as last time since i did not commit the bezier-based implementation yet and other changes are jsut code structure changes. I am not completely certain why it looks so bad on lower values, the aspect ratio might part of the problem.

I really want to get this implementation right since this would probably be the most accurate real-time bloom implementation out of all major game engines/frameworks.

I'll create an interactive explainer of the bezier-curve-controlled bloom when this is done but i promise there are physics behind the bezier curves. Here are the sketches i've been doing when coming up with it:
Screenshot from 2022-12-02 18-14-49
(The blobs are light distributions)

> Perhaps we can do the blur pyramid multiple times (dependent on resolution)?

Those artifacts only occur horizontally so it might improve once the aspect ratio bug is fixed.

> While you're working on bloom, do you mind adding the fix from https://github.com/bevyengine/bevy/pull/6802?

I'll try... I hope no crazy changes there

**Comment by StarLederer** [comment]
This took me embarassingly long to complete but i made the visualizer.

Light enters from the bottom-center and travels upwards; in the current blur pyramid setup the light scatters in every direction between -90deg and 90deg angles; the blobs/lines represent the end blend factor of each mip (they always add up to 1 because no light is created and no light lost).

Here are some results that i got by manually typing in some configurations:

This is what a setup which i normally use in Unity (all levels bend with 0.8 intensity except the last one, which uses 0.3) looks like
common
I like this one because it creates what essentially is the opposite of "the ugly ring", however, I see that for some reason one of the blend levels at strong angles caves in for some reason. I have no clue why this happens but unless i made some mistake, that is what constant blend levels (except final one) look like. *There is a chance it is an error in my algorithm, but it seems to work overall and blend factors do add up to 1 so it probably is just a weird math/physics thing that no one ever visualized.*

This one is what this PR currently looks like (everything blended with 0.3)
current
This creates the "forward scattering" effect usually desired for cloud and mist rendering but does not make for convincing lens scattering. Also, this one is the "ugly ring" one.

And this is what I think the best looking bloom would look like
desired
I noticed that everyone really likes smooth and long falloffs, and intuitively this is what I guess will create it. Also this one is inspired by the first graph, but i made the final layer blend at 0.5 which should create a very bloomy, but not blurry effect because most of the bloom is going to be the smallest mips.

This is what the curves for each look like (same order as above):

common (curve)

current (curve)

desired (curve)

My current plan is to make some editable curves, plug them into this PR, and see how that looks. After that I'll probably have to go back to the original idea of mathematically representing this curve because I realized that what I'm making rn will not work with textures which we do need to create lens dirt or ui blur effects.

Would much appreciate some help with the mathematical representation because as I said before I'm not that good at it. If you know anyone or a place i can ask, please let me know. What I'm looking for seems to be something like this:
image
(function should be able to recreate any of these curves)

**Comment by StarLederer** [comment]
Just figured thiese two might be interesting too.

This PR at intensity of 0.5 (the point at which it looks pretty good but too blurry):


This PR at intensity 0.9 (very blurry):

**Comment by StarLederer** [comment]
I improved the visualizer, made it interactive (very badly atm), and published it to GH pages [here](https://starlederer.github.io/bloom/)

**Comment by JMS55** [comment]
The visualizer is cool! Looking forward to seeing what you come up with.

**Comment by StarLederer** [comment]
The commit above adds a dirty implementation of the blending function. I need to rename some variables and add comments but the function already produces intended values and the bloom looks great (imo). The customizability  goals also seem to have been met.

Proposed default look:
proposed default

Extra soft bloom:
slight bloom

"Ugly ring" is still possible (if someone wants it):
ugly ring

Screen blur is limited in customizability but should satisfy most designers:
screen blur
screen blur 2

# Up next

Tidy things up and implement the fix you mentioned. After that PR should be ready to merge.

**Comment by JMS55** [comment]
Looks great! I'll need a few days before I can try it out and review the code, currently in the middle of final exams.

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/mod.rs]: This code could use an explanation, a link to a desmos graph, or similar comment explaining it to the best of your ability.

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/mod.rs]: The solution to this is to use pipeline specialization to only create uniforms + uniform binding for when bloom settings are configured to need it, and then put the computations in the shader behind an #ifdef. You can see an example of this here https://github.com/JMS55/bevy/blob/ade10a58d9db19c786336a358d0425e354a75e65/crates/bevy_pbr/src/taa/mod.rs. Let me know if you have any questions on how to do this.

File: crates/bevy_core_pipeline/src/bloom/bloom.wgsl
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/bloom.wgsl:10]: I can't leave a comment on an unchanged piece of code, but the comment at the top of this file still mention additively blending. Need to change that to just blend.

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/settings.rs]: Types implementing Reflect need to be registered with the app in the plugin build() function. Same with PrefilterSettings.

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/settings.rs]:
```suggestion
    pub fn old_school() -> Self { Self {
```

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/settings.rs]:
```suggestion
/// to an SDR image or when emulating the 1990s-2000s game look.
```

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/settings.rs:106]: Right now bloom will not work without HDR rendering. The bloom extract function skips SDR cameras. If you want to support SDR, that needs changing.

File: examples/3d/bloom.rs
Comment (JMS55) [inline examples/3d/bloom.rs:39]: Related to the above SDR comment, should probably change the wording of "HDR must be enabled" to "HDR is highly recommended for natual looking bloom" or something similar.

File: examples/3d/bloom.rs
Comment (JMS55) [inline examples/3d/bloom.rs]: Please remove these comments. If you really want to separate these two blocks more, split this into two different functions/systems.

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/settings.rs]: Same comment about HDR/SDR applies here.

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/settings.rs]:
```suggestion
```

Can probably remove this line, or combine it with the doc comments for this struct.

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/settings.rs:38]: Move intensity to be the first struct field - it's the most important imo.

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/settings.rs]: Each of these fields needs an explanation about how they effect the image (to the best that you can), as well as a comment explaining the valid range, and that outside values will be clamped (as PrefilterSettings does).

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/mod.rs]: Should probably clamp these values to between [0, 1] before using them, like the example demo does.

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/settings.rs]:
```suggestion
pub struct BloomPrefilterSettings {
```

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/settings.rs]:
```suggestion
pub enum BloomCompositeMode {
```

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/settings.rs]: What's this one do? Could use a doc comment.

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/settings.rs:21]: Change the order of structs in this file to BloomSettings -> PrefilterSettings -> BloomMode please.

File: examples/3d/bloom.rs
Comment (JMS55) [inline examples/3d/bloom.rs]: Rather than expose these "advanced" settings in the demo, I'd rather only keep intensity, and have a toggle between the different natural/old_school+sdr/screen_blur/off presets,. and maybe one with a preconfigured filter as well.

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/bloom/mod.rs]: I'm still not convinced we want to go down to 1x1 textures - that's super small and seems unnecessary. 4x4 should be the limit imo. We can leave it if you think it's worthwhile, though.

**Review by JMS55 — CHANGES_REQUESTED** [review]
Turns out I had time today actually. Visually looks super good to me, really nice job. I'm very happy with this approach. Below are my review requests.

---

Pipeline specialization will be needed to clean up most of your TODOs. Let me know if you run into any confusion, I'd be happy to explain it.

Some comments about SDR support, easy to address.

https://github.com/bevyengine/bevy/pull/6802 we should get this merged as part of this PR, along with double checking that transparent windows still work. Transparent windows were broken with bevy 0.9's bloom, but fixed in my original PR. We should test that this PR keeps that fix still.

Mostly doc improvements are needed, besides the above.

---

Once all the above is addressed, and we go through a final round of review, I'd be happy to merge this to my open PR to bevy, and then we can get it merged into the main repo :). Thanks again for working on this!

Other thing I'd like to get done for bloom this cycle is WebGL2 fallback support, using separate textures instead of mips. I'll work on that as a later PR through.

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (StarLederer) [reply crates/bevy_core_pipeline/src/bloom/settings.rs:106]: I know, was trying to leave that up to someone else to change but i guess wrote misleading comments that assume that is already possible. This sounds like a simple change so i should probably actually make it

File: examples/3d/bloom.rs
Comment (StarLederer) [reply examples/3d/bloom.rs]: Oops, that was never supposed to make it into the PR

File: examples/3d/bloom.rs
Comment (StarLederer) [reply examples/3d/bloom.rs]: I considered that, but decided against it because that makes it harder to test the effect and i didn't want to ruin the experience for whoever modifies it after us. On top of that, i revamped the blend function (again) and now the values make more sense. Are you sure we should change this to intensity + presets?

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (StarLederer) [reply crates/bevy_core_pipeline/src/bloom/settings.rs]: I swear there was a doc comment before... either way, it's back now

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (StarLederer) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: The values are clamped in the demo.

Should they be clamped here as well? I could implement a validate function for BloomSettings that returns a result of () or an error describing why some values are invalid. Could also add a strictness parameter that would make the function treat using BloomCompositeMode::EnergyConserving together with non-0 threshold as an invalid configuration as well. This could be pretty useful for GUIs people might build in the future

**Comment by StarLederer** [comment]
> There is a chance it is an error in my algorithm

And so it was... It is fixed now and [my visualizer](https://starlederer.github.io/bloom/) does not produce the weird spike at the lowest frequency mip blend level anymore. I also revamped the blend function (again 😭) in accordance with the new results which made its parameters much more intuitive to read and improved control over screen blur.

The visual results are pretty similar (depending on the configuration) so I shall not flood this conversation with screenshots this time.

**Comment by StarLederer** [comment]
Also implemented most small fixes outlined in the review and marked conversations as resolved. There are still a couple i need to address, will get to them as soon as I can

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (StarLederer) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: Added a doc comment and a link to the Desmos graph. Would have preferred to link to my visualizer instead but it does not have any explanations yet. Might end up adding a blog style explanation to it at some point, will PR a link then.

I'm not resolving this conversation, because i can't validate whether my own explanation is good.

File: crates/bevy_core_pipeline/src/bloom/settings.rs
Comment (StarLederer) [reply crates/bevy_core_pipeline/src/bloom/settings.rs:106]: So it wasn't a simple change after all because but it's done now.

It required to specialize the upsampling pipeline, but I now know how to specialize pipelines which is great (thanks to your example). Also the pipeline specialization will help move blend factor computation to the shader so that's also great.

Please resolve the conversation if everything works for you

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (StarLederer) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: I wonder if the added complexity is worth it and weather the performance we save here is more than we loose by specializing the pipeline (idk if pipeline specialization is costly)

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (superdump) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: Using really low res highest level mips introduced visual artifacts in the original implementation.

Comment (JMS55) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: I believe I tested the new version on bistro, and it was fine with 1x1. We should double check again, though.

File: examples/3d/bloom.rs
Comment (JMS55) [reply examples/3d/bloom.rs]: Up to you on what you think is best. Leaving it as is is also fine. We can always wait and see what other people think when this gets reviewed on the main bevy repo (sorry, didn't see this comment until now).

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (JMS55) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: Probably worth it. The specialization is not too expensive. It's basically just compiling different variations on a shader at runtime based on the key, which is then cached in subsequent frames. Easily hideable in a loading screen when first setting up the camera.

**Comment by JMS55** [comment]
Happy near year. I wanted to let you know that I will be somewhat away for the next ~2 weeks. I'll have internet access, but somewhat limited. 

This PR is almost ready. As I'll be away, you should take over the bloom PR (assuming you have time). I've invited you as a "collaborator" to this repository (please don't modify other branches besides the bloom one). Feel free to merge this once you feel it's ready, and then update the PR description on the bevy repo so that people can start reviewing it before the next release :)

**Comment by StarLederer** [comment]
Happy new year! Thanks for assisting me in this PR so far. I've accepted the invitation and will treat this fork responsibly

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (StarLederer) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: Ok done, marking as resolves since you are on a holiday

File: examples/3d/bloom.rs
Comment (StarLederer) [reply examples/3d/bloom.rs]: I think I'll leave it

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (StarLederer) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: I added a slight limit but not going to 1x1 means out maximum scattering angel is not 90deg. Oh well whatever, it wasn't really 1x1 anyway unless you had a perfectly square window

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (StarLederer) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: I think the validation is overkill and clamping the values in contribution calculation adds overhead. I'd rather just have users set correct values based on doc comments and ideally have a UI for this at some point that will not go beyond valid range. On top of that some values should only be clamped in physically based mode which adds some inconsistency. I don't want to add these clamps, but if we find them necessary we can throw them in in a future PR

**Comment by StarLederer** [comment]
Merged upstream.

Tried implementing #6802 first but it used newer code that was very different in some parts. I'm surprised merging was not too bad considering the change with uniform extraction. #6802 should be really easy to implement now

**Comment by StarLederer** [comment]
Done!

Not sure what's the expected behavior for bloom on a transparent window but I am getting these results:

*with custom viewport*
Screenshot from 2023-01-03 23-44-46

*with default viewport*
Screenshot from 2023-01-03 23-45-21

**Comment by StarLederer** [comment]
All done! 🎉🎉🎉

I all planned changes have been implemented, change requests addressed and #6802 implemented. Transparent windows and custom view ports both work. I will now proceed to merge this and update the PR on upstream.

**Pull request merged**
